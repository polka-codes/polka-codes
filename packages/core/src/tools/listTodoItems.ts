import { z } from 'zod'
import { type FullToolInfo, type ToolHandler, type ToolInfo, ToolResponseType } from '../tool'
import type { TodoProvider } from './provider'

export const toolInfo = {
  name: 'listTodoItems',
  description: 'List all to-do items. If an id is provided, it lists all sub-items for that id.',
  parameters: z.object({
    id: z.string().nullish(),
  }),
} as const satisfies ToolInfo

export const handler: ToolHandler<typeof toolInfo, TodoProvider> = async (provider, args) => {
  if (!provider.listTodoItems) {
    return {
      type: ToolResponseType.Error,
      message: {
        type: 'error-text',
        value: 'Not possible to list to-do items.',
      },
    }
  }
  const { id } = toolInfo.parameters.parse(args)
  const items = await provider.listTodoItems(id)
  return {
    type: ToolResponseType.Reply,
    message: {
      type: 'json',
      value: items,
    },
  }
}

export default {
  ...toolInfo,
  handler,
} satisfies FullToolInfo
