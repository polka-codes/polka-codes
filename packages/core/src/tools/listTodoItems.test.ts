import { describe, expect, it, spyOn } from 'bun:test'
import listTodoItems from './listTodoItems'
import type { TodoProvider } from './provider'
import type { TodoItem } from './todo'

describe('listTodoItems', () => {
  const mockTodos: TodoItem[] = [
    { id: '10', title: 'item 10', description: '', status: 'open' },
    { id: '2', title: 'item 2', description: '', status: 'completed' },
    { id: '1.2', title: 'sub-item 1.2', description: '', status: 'completed' },
    { id: '1.1', title: 'sub-item 1.1', description: '', status: 'open' },
    { id: '1', title: 'item 1', description: '', status: 'open' },
  ]

  const createMockProvider = (items: TodoItem[]): TodoProvider => ({
    listTodoItems: async (id?: string | null, status?: string | null) => {
      let resultItems = id ? items.filter((item) => item.id.startsWith(`${id}.`)) : items

      if (status) {
        resultItems = resultItems.filter((item) => item.status === status)
      }

      resultItems.sort((a, b) => {
        const aParts = a.id.split('.')
        const bParts = b.id.split('.')
        const len = Math.min(aParts.length, bParts.length)
        for (let i = 0; i < len; i++) {
          const comparison = aParts[i].localeCompare(bParts[i], undefined, { numeric: true })
          if (comparison !== 0) {
            return comparison
          }
        }
        return aParts.length - bParts.length
      })

      return resultItems
    },
    getTodoItem: () => {
      throw new Error('Function not implemented.')
    },
    updateTodoItem: () => {
      throw new Error('Function not implemented.')
    },
  })

  it('should list all todo items sorted correctly', async () => {
    const provider = createMockProvider(mockTodos)
    const result = await listTodoItems.handler(provider, {})
    expect(result).toMatchSnapshot()
  })

  it('should filter items by status', async () => {
    const provider = createMockProvider(mockTodos)
    const result = await listTodoItems.handler(provider, { status: 'completed' })
    expect(result).toMatchSnapshot()
  })

  it('should list sub-items for a given id', async () => {
    const provider = createMockProvider(mockTodos)
    const listSpy = spyOn(provider, 'listTodoItems')
    const result = await listTodoItems.handler(provider, { id: '1' })
    expect(listSpy).toHaveBeenCalledWith('1', undefined)
    expect(result).toMatchSnapshot()
  })
})
