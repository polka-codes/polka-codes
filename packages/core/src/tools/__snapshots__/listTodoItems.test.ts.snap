// Bun Snapshot v1, https://bun.sh/docs/test/snapshots

exports[`listTodoItems should list all todo items sorted correctly 1`] = `
{
  "message": {
    "type": "json",
    "value": [
      {
        "description": "",
        "id": "1",
        "status": "open",
        "title": "item 1",
      },
      {
        "description": "",
        "id": "1.1",
        "status": "open",
        "title": "sub-item 1.1",
      },
      {
        "description": "",
        "id": "1.2",
        "status": "completed",
        "title": "sub-item 1.2",
      },
      {
        "description": "",
        "id": "2",
        "status": "completed",
        "title": "item 2",
      },
      {
        "description": "",
        "id": "10",
        "status": "open",
        "title": "item 10",
      },
    ],
  },
  "type": "Reply",
}
`;

exports[`listTodoItems should filter items by status 1`] = `
{
  "message": {
    "type": "json",
    "value": [
      {
        "description": "",
        "id": "1.2",
        "status": "completed",
        "title": "sub-item 1.2",
      },
      {
        "description": "",
        "id": "2",
        "status": "completed",
        "title": "item 2",
      },
    ],
  },
  "type": "Reply",
}
`;

exports[`listTodoItems should list sub-items for a given id 1`] = `
{
  "message": {
    "type": "json",
    "value": [
      {
        "description": "",
        "id": "1.1",
        "status": "open",
        "title": "sub-item 1.1",
      },
      {
        "description": "",
        "id": "1.2",
        "status": "completed",
        "title": "sub-item 1.2",
      },
    ],
  },
  "type": "Reply",
}
`;
